src/PromisifiedFS.js

```diff
@@ -0,0 +1,181 @@
+  // strip out callbacks
+  if (typeof opts === "undefined" || typeof opts === "function") {
+    opts = {};
+  }
+  // expand string options to encoding options
+  if (typeof opts === "string") {
+    opts = {
+      encoding: opts,
+    };
+  }
+  return [filepath, opts];
+}
+
+function cleanParams2(oldFilepath, newFilepath) {
+  // normalize paths
+  oldFilepath = path.normalize(oldFilepath);
+  newFilepath = path.normalize(newFilepath);
+  return [oldFilepath, newFilepath];
+}
+
+module.exports = class PromisifiedFS {
+  constructor(name, { wipe, url } = {}) {
+    this._idb = new IdbBackend(name);
+    this._cache = new CacheFS(name);
+    this._opts = { wipe, url };
+    if (url) {
+      this._http = new HttpBackend(url)
+    }
+    this._initPromise = this._init()
+    // YES THIS IS A MESS
+    this.readFile = this._wrap(this._readFile.bind(this))
+    this.writeFile = this._wrap(this._writeFile.bind(this))
+    this.unlink = this._wrap(this._unlink.bind(this))
+    this.readdir = this._wrap(this._readdir.bind(this))
+    this.mkdir = this._wrap(this._mkdir.bind(this))
+    this.rmdir = this._wrap(this._rmdir.bind(this))
+    this.rename = this._wrap(this._rename.bind(this))
+    this.stat = this._wrap(this._stat.bind(this))
+    this.lstat = this._wrap(this._lstat.bind(this))
+    // Needed so things don't break if you destructure fs and pass individual functions around
+    this.readFile = this.readFile.bind(this)
+    this.writeFile = this.writeFile.bind(this)
+    this.unlink = this.unlink.bind(this)
+    this.readdir = this.readdir.bind(this)
+    this.mkdir = this.mkdir.bind(this)
+    this.rmdir = this.rmdir.bind(this)
+    this.rename = this.rename.bind(this)
+    this.stat = this.stat.bind(this)
+    this.lstat = this.lstat.bind(this)
+    this.readlink = this.readlink.bind(this)
+    this.symlink = this.symlink.bind(this)
+  }
+  async _init() {
+    if (this._initPromise) return this._initPromise
+    if (this._opts.wipe) {
+      await this._wipe();
+    } else {
+      await this._loadSuperblock();
+    }
+    await this._saveSuperblock();
+  }
+  async _wipe() {
+    await this._idb.wipe()
+    if (this._http) {
+      const text = await this._http.fetchSuperblock()
+      if (text) {
+        this._cache.loadSuperBlock(text)
+      }
+    }
+  }
+  _saveSuperblock() {
+    return this._idb.storeSuperblock(this._cache._root);
+  }
+  async _loadSuperblock() {
+    let root = await this._idb.fetchSuperblock()
+    if (!root && this._http) {
+      root = await this._http.fetchSuperblock()
+    }
+    if (root) {
+      this._cache.loadSuperBlock(root);
+    }
+  }
+  _wrap (fn) {
+    return async (...args) => {
+      await this._init()
+      await this._loadSuperblock()
+      let data, err
+      try {
+        data = await fn(...args)
+      } catch (e) {
+        err = e
+      }
+      await this._saveSuperblock()
+      if (err) throw err
+      return data
+    }
+  }
+  async _readFile(filepath, opts) {
```

Comment:
From a user perspective this makes sense... it makes me feel icky though. Because if you're reading the code, it says it's `readFile` but it's _not_ actually `readFile`.

Say, I'm pretty sure this kind of meta-programming is a good use case for decorators. You wouldn't know how to convert `_wrap` to a decorator would you?